### PR TITLE
[inductor] Add prims._inductor_bucketize and add lowerings

### PR DESCRIPTION
**Pull request by davidberard98**
**TL;DR**: This PR is a first step in adding lowerings for torch.bucketize. It adds an initial lowering for this op - but because this  implementation is not currently efficient, it registers the lowering for prims._inductor_bucketize. After we make the implementation more efficient, we'll remove prims._inductor_bucketize and add the lowering directly to torch.bucketize.

**Background - torch.bucketize**: torch.bucketize(values, boundaries, right=False): for an arbitrary tensor of values and a non-decreasing 1D tensor of boundaries that define buckets, it returns the index of the bucket that each of the values will fall in. e.g. for values [0, 1, 2, 3, 4] and boundaries [1, 3], it will return [0, 0, 1, 1, 2].

**Implementation**: This PR adds a new inductor op called "bucketize". In this PR it only has a triton implementation - for CPU it is a fallback. The triton implementation uses a binary search in `triton_helpers.py`. This PR also adds a new prim `_inductor_bucketize()` for testing purposes and adds lowering for this op.

~~**"right"**: The current behavior of the "right" kwarg in the inductor op is the opposite of the behavior of the torch op. "right" controls how the op treats a value that is equal to one of the boundary values. In the torch op, "right=True" means "if a value is equal to a boundary value, then put it in the bucket to the right". In the inductor op, "right=True" means "the right boundary of a bucket is closed". These are opposite. **I'm open to switching the behavior of the inductor op** - but I chose to implement this way because I think it makes more sense, and I think the torch.bucketize behavior may have been a mistake (it's the opposite of numpy.digitize).~~ Switched the behavior of the inductor bucketize op to match the torch op

* places where "right" means "if a value is equal to a boundary value, then put it in the bucket to the right" (i.e. current torch.bucketize behavior)
  + current torch.bucketize behavior
  + table in [torch.bucketize docs](https://pytorch.org/docs/stable/generated/torch.bucketize.html)
* places where "right" means "the right boundary of a bucket is closed":
  + the text description of [torch.bucketize docs](https://pytorch.org/docs/stable/generated/torch.bucketize.html) (observed in #91580)
  + [numpy.digitize](https://numpy.org/doc/stable/reference/generated/numpy.digitize.html) (which is basically the same op)

**Performance**: Benchmark script: "values" as a [16, 1024, 1024] float32 tensor and "boundaries" as a [1025] tensor (i.e. defining 1024 buckets).

As is:
```
Eager 0.30117499828338623 ms
PT2   0.9298200011253357 ms
```

But performance improves significantly if we add an additional pointwise autotuning config (WIP in #104456):
```
Eager 0.3015420138835907 ms
PT2   0.23028500378131866 ms
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -833,6 +841,35 @@ def reduction(name, dtype, src_dtype, reduction_type, index, value):
...
+                right: bool,
+            ):
+                """
+                [Note: Inductor bucketize op]
+
+                Given values (tensor) and offsets_name (reference to the name of a 1D
+                tensor), calculate the bucket that each value belongs to.
+
+                e.g. for values [-1, 0, 1, 2, 3, 4, 5, 9], offsets [0, 4, 8], right=False
+                return =        [ 0, 1, 1, 1, 1, 2, 2, 3].
```
Comment (aakhundov) [inline torch/_inductor/codegen/common.py:859]: Although this, indeed, seems to correspond to the semantics of `torch.bucketize`, in the jagged tensor context, seems that we'll need to subtract 1 from the result (as the values falling into the first bucket should have index 0).

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -1551,6 +1551,49 @@ def inner_fn(index):
     )
 
 
+@register_lowering(inductor_prims._bucketize, type_promotion_kind=None)
```
Comment (aakhundov) [inline torch/_inductor/lowering.py:1554]: The [documentation](https://pytorch.org/docs/stable/generated/torch.bucketize.html) for `torch.bucketize` specifies that the `boundaries` "must contain a strictly increasing sequence, or the return value is undefined". In the jagged tensor context, the buckets can be zero-size (corresponding to empty sequences in the jagged tensor). I guess, our lowering here should support zero-sized buckets?

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1301,6 +1323,14 @@ def reduction_resize(self, value):
         sizes[-1] = "None"
         return f"{value}[{', '.join(sizes)}]"
 
+    def reduction_size_str(self):
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/_inductor/codegen/triton.py:1326]: This is leftover from a bad rebase resolution I think, will remove it

File: torch/_inductor/triton_helpers.py [hunk 4]
```diff
@@ -117,3 +117,51 @@ def _any_combine(a, b):
...
+
+
+@triton.jit
+def bucketize_binary_search(
+    values,  # 1D tensor
+    offsets_ptr,
+    indexing_dtype,
+    right,  # bool: if true, use intervals closed on the right
+    OFFSETS_SIZE: int,
+    BLOCK_SHAPE,
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/_inductor/triton_helpers.py:129]: It's actually a list/tuple. IDK if it's always 1D or if it can be multi-dimensional sometimes, so I was planning to just pass it in directly like this.

Added a comment to clarify

File: torch/_inductor/codegen/common.py (hunk 1 above)
Comment (davidberard98) [reply torch/_inductor/codegen/common.py]: In some rough benchmarks I didn't see any measurable difference in perf from subtracting the 1 vs. not having to do that (in fbgemm lowerings)

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (davidberard98) [reply torch/_inductor/lowering.py:1554]: Good point, updated the docstring in common.py. In practice it seems like torch.bucketize obeys this behavior and we can define the behavior to work for jagged tensors.

File: torch/_inductor/lowering.py [hunk 5]
```diff
@@ -1556,6 +1556,49 @@ def inner_fn(index):
...
+def _inductor_bucketize(
+    input: TensorBox,
+    boundaries: TensorBox,
+    *,
+    out_int32: bool = False,
+    right: bool = False,
+):
+    assert len(boundaries.get_size()) == 1
+
+    if input.get_device().type != "cuda" or boundaries.get_device().type != "cuda":
```
Comment (jansel) [inline torch/_inductor/lowering.py:1569]: Use the is_triton() helper (which does the same thing)

File: torch/_inductor/lowering.py [hunk 6]
```diff
@@ -1556,6 +1556,49 @@ def inner_fn(index):
...
+            input, boundaries, out_int32=out_int32, right=right
+        )
+
+    boundaries_size = boundaries.get_size()[0]
+    boundaries_loader = boundaries.make_loader()
+    device = input.get_device()
+    input_loader = input.make_loader()
+
+    index_dtype = torch.int32 if out_int32 else torch.int64
+    triton_dtype = "tl.int32" if out_int32 else "tl.int64"
```
Comment (jansel) [inline torch/_inductor/lowering.py:1580]: Let's not expose the triton_dytpe in the device-agnostic IR.  This will make it harder to extend to non-triton backends.

File: torch/_inductor/lowering.py [hunk 7]
```diff
@@ -1556,6 +1556,49 @@ def inner_fn(index):
...
+
+    def inner_fn(index):
+        val = input_loader(index)
+        indices = ops.bucketize(
+            val,
+            boundaries.get_name(),
+            ops.index_expr(boundaries_size, index_dtype),
+            triton_dtype,
+            not right,  # ops.bucketize and torch.bucketize have opposite semantics for "right"
```
Comment (jansel) [inline torch/_inductor/lowering.py:1589]: My initial feelings here are we should match the torch semantics rather than the numpy one.  That will at least make our codebase self-consistent.  I don't feel that strongly about that though.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/triton_helpers.py [hunk 8]
```diff
@@ -117,3 +117,37 @@ def _any_combine(a, b):
...
 def any(a, dim):
     return tl.reduce(a, dim, _any_combine)
+
+
+@triton.jit
+def bucketize_binary_search(
+    values,  # 1D tensor
+    offsets_ptr,
+    indexing_dtype,
+    right,  # bool: if true, use intervals closed on the right
```
Comment (jansel) [inline torch/_inductor/triton_helpers.py:127]: outdated comment?

Comment (davidberard98) [reply torch/_inductor/triton_helpers.py:127]: whoops, thanks for catching that

File: torch/_inductor/inductor_prims.py [hunk 9]
```diff
@@ -71,3 +72,24 @@ def eager_force_stride(input_tensor, stride):
...
+def _inductor_bucketize_meta(input, boundaries, *, out_int32=False, right=False):
+    return torch.empty_like(
+        input,
+        memory_format=torch.preserve_format,
+        dtype=(torch.int32 if out_int32 else torch.int64),
+    )
+
+
+_bucketize = _prims._make_prim(
+    schema="_inductor_bucketize(Tensor input, Tensor boundaries, *, bool out_int32=False, bool right=False) -> Tensor",
```
Comment (peterbell10) [inline torch/_inductor/inductor_prims.py:90]: What's the advantage of adding a prim rather than just not decomposing `aten.bucketize` in inductor?

Comment (davidberard98) [reply torch/_inductor/inductor_prims.py:90]: because this implementation is not currently efficient, I'm the lowering for prims._inductor_bucketize so that it doesn't affect perf for models that use torch.bucketize. After we make the implementation more efficient, we'll remove prims._inductor_bucketize and add the lowering directly to torch.bucketize.

Comment (peterbell10) [reply torch/_inductor/inductor_prims.py:90]: Oh sorry for the noise, should have read the PR description more carefully.

**Review by jansel — APPROVED** [review]
